Migrating partitioned Hive tables to Iceberg fails on Glue due to SkipArchiveRequestHandler

```
Query 20240404_083846_00024_8h73f failed: Failed to migrate table
io.trino.spi.TrinoException: Failed to migrate table
        at io.trino.plugin.iceberg.procedure.MigrateProcedure.doMigrate(MigrateProcedure.java:262)
        at io.trino.plugin.iceberg.procedure.MigrateProcedure.migrate(MigrateProcedure.java:178)
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:732)
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:757)
        at io.trino.execution.CallTask.execute(CallTask.java:211)
        at io.trino.execution.CallTask.execute(CallTask.java:71)
        at io.trino.execution.DataDefinitionExecution.start(DataDefinitionExecution.java:146)
        at io.trino.execution.SqlQueryManager.createQuery(SqlQueryManager.java:251)
        at io.trino.dispatcher.LocalDispatchQuery.lambda$startExecution$7(LocalDispatchQuery.java:143)
        at io.trino.$gen.Trino_414____20240325_034800_2.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: io.trino.spi.TrinoException: Failed to fetch partitions from Glue Data Catalog
        at io.trino.plugin.hive.metastore.glue.GlueHiveMetastore.getPartitionValues(GlueHiveMetastore.java:840)
        at io.trino.plugin.hive.metastore.glue.GlueHiveMetastore.getPartitionNamesByFilter(GlueHiveMetastore.java:810)
        at io.trino.plugin.iceberg.procedure.MigrateProcedure.listAllPartitions(MigrateProcedure.java:304)
        at io.trino.plugin.iceberg.procedure.MigrateProcedure.doMigrate(MigrateProcedure.java:222)
        ... 12 more
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Unsupported request: {DatabaseName: schema1,TableName: table1,Expression: ,Segment: {SegmentNumber: 0,TotalSegments: 5},MaxResults: 1000,ExcludeColumnSchema: true,}
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
        at io.trino.plugin.hive.metastore.glue.GlueHiveMetastore.getPartitionValues(GlueHiveMetastore.java:831)
        ... 15 more
Caused by: java.lang.IllegalArgumentException: Unsupported request: {DatabaseName: schema1,TableName: table1,Expression: ,Segment: {SegmentNumber: 0,TotalSegments: 5},MaxResults: 1000,ExcludeColumnSchema: true,}
        at io.trino.plugin.iceberg.catalog.glue.SkipArchiveRequestHandler.beforeExecution(SkipArchiveRequestHandler.java:54)
        at com.amazonaws.AmazonWebServiceClient.beforeClientExecution(AmazonWebServiceClient.java:787)
        at com.amazonaws.services.glue.AWSGlueClient.getPartitions(AWSGlueClient.java:7065)
        at io.trino.plugin.hive.metastore.glue.AwsSdkUtil$1.lambda$computeNext$0(AwsSdkUtil.java:62)
        at io.trino.plugin.hive.aws.AwsApiCallStats.call(AwsApiCallStats.java:37)
        at io.trino.plugin.hive.metastore.glue.AwsSdkUtil$1.computeNext(AwsSdkUtil.java:62)
        at com.google.common.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:146)
        at com.google.common.collect.AbstractIterator.hasNext(AbstractIterator.java:141)
        at java.base/java.util.Iterator.forEachRemaining(Iterator.java:132)
        at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
        at io.trino.plugin.hive.metastore.glue.GlueHiveMetastore.getPartitionValues(GlueHiveMetastore.java:872)
        at io.trino.plugin.hive.metastore.glue.GlueHiveMetastore.lambda$getPartitionValues$29(GlueHiveMetastore.java:827)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at io.airlift.concurrent.BoundedExecutor.drainQueue(BoundedExecutor.java:80)
        ... 3 more
```
https://trinodb.slack.com/archives/C0305TQ05KL/p1712220221408549?thread_ts=1712204419.522639&cid=C0305TQ05KL

--- a/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/SkipArchiveRequestHandler.java
+++ b/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/SkipArchiveRequestHandler.java
@@ -15,12 +15,14 @@ package io.trino.plugin.iceberg.catalog.glue;
 
 import com.amazonaws.AmazonWebServiceRequest;
 import com.amazonaws.handlers.RequestHandler2;
+import com.amazonaws.services.glue.model.BatchGetPartitionRequest;
 import com.amazonaws.services.glue.model.CreateDatabaseRequest;
 import com.amazonaws.services.glue.model.CreateTableRequest;
 import com.amazonaws.services.glue.model.DeleteDatabaseRequest;
 import com.amazonaws.services.glue.model.DeleteTableRequest;
 import com.amazonaws.services.glue.model.GetDatabaseRequest;
 import com.amazonaws.services.glue.model.GetDatabasesRequest;
+import com.amazonaws.services.glue.model.GetPartitionsRequest;
 import com.amazonaws.services.glue.model.GetTableRequest;
 import com.amazonaws.services.glue.model.GetTablesRequest;
 import com.amazonaws.services.glue.model.UpdateTableRequest;
@@ -41,7 +43,10 @@ public class SkipArchiveRequestHandler
                 request instanceof CreateTableRequest ||
                 request instanceof DeleteTableRequest ||
                 request instanceof GetTablesRequest ||
-                request instanceof GetTableRequest) {
+                request instanceof GetTableRequest ||
+                // The following requests are required for migrate procedure
+                request instanceof GetPartitionsRequest ||
+                request instanceof BatchGetPartitionRequest) {
             return request;
         }
         throw new IllegalArgumentException("Unsupported request: " + request);

--- a/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseSharedMetastoreTest.java
+++ b/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseSharedMetastoreTest.java
@@ -168,6 +168,38 @@ public abstract class BaseSharedMetastoreTest
         }
     }
 
+    @Test
+    public void testMigrateTable()
+    {
+        String tableName = "test_migrate_" + randomNameSuffix();
+        String hiveTableName = "hive.%s.%s".formatted(schema, tableName);
+        String icebergTableName = "iceberg.%s.%s".formatted(schema, tableName);
+
+        assertUpdate("CREATE TABLE " + hiveTableName + " AS SELECT 1 id", 1);
+        assertQueryFails("SELECT * FROM " + icebergTableName, "Not an Iceberg table: .*");
+
+        assertUpdate("CALL iceberg.system.migrate('" + schema + "', '" + tableName + "')");
+        assertQuery("SELECT * FROM " + icebergTableName, "VALUES 1");
+
+        assertUpdate("DROP TABLE " + tableName);
+    }
+
+    @Test
+    public void testMigratePartitionedTable()
+    {
+        String tableName = "test_migrate_" + randomNameSuffix();
+        String hiveTableName = "hive.%s.%s".formatted(schema, tableName);
+        String icebergTableName = "iceberg.%s.%s".formatted(schema, tableName);
+
+        assertUpdate("CREATE TABLE " + hiveTableName + " WITH (partitioned_by = ARRAY['part']) AS SELECT 1 id, 'test' part", 1);
+        assertQueryFails("SELECT * FROM " + icebergTableName, "Not an Iceberg table: .*");
+
+        assertUpdate("CALL iceberg.system.migrate('" + schema + "', '" + tableName + "')");
+        assertQuery("SELECT * FROM " + icebergTableName, "VALUES (1, 'test')");
+
+        assertUpdate("DROP TABLE " + tableName);
+    }
+
     private long getLatestSnapshotId(String schema)
     {
         return (long) computeScalar(format("SELECT snapshot_id FROM iceberg.%s.\"nation_test$snapshots\" ORDER BY committed_at DESC FETCH FIRST 1 ROW WITH TIES", schema));

